docs/features/auth/certificate-authentication.md

```diff
@@ -0,0 +1,99 @@
+---
+title: "Authentication with certificate via ASP.NET Core authentication filters"
+layout: default
+---
+
+The `Arcus.WebApi.Security` package provides a mechanism that uses the client certificate of the request to grant access to a web application.
+This authentication process consists of following parts:
+
+1. Find the client certificate configured on the HTTP request
+2. Determine which properties of the received client certificate are used for authentication
+3. The property value(s) of the client certificate matches the value(s) determined via configured secret provider, configuration or custom implementation
+
+The package allows two ways to configure this type of authentication mechanism in an <span>ASP.NET</span> application:
+- [Globally enforce certificate authentication](#Globally-enforce-certificate-authentication)
+- [Enforce certificate authentication per controller or operation](#Enforce-certificate-authentication-per-controller-or-operation)
+
+## Globally enforce certificate authentication
+
+### Introduction
+
+The `CertificateAuthenticationFilter` can be added to the request filters in an <span>ASP.NET</span> Core application.
+This filter will then add authentication to all endpoints via one or many certificate properties configurable on the filter itself.
+
+### Usage
+
+The authentication requires a service dependency to be registered with the services container of the <span>ASP.NET</span> request pipeline, which can be one of the following:
+- `ICachedSecretProvider` or `ISecretProvider`: built-in or you implementation of the secret provider
+- `Configuration`: key/value pairs in the configuration of the <span>ASP.NET</span> application
+- `IX509ValidationLocation`: custom implementation that implements the validation location interface
+
+This registration of the service is typically done in the `ConfigureServices` method of the `Startup` class.
+
+Each certificate property that should be validated can use a different service dependency. 
+This mapping of what service which property uses, is defined in an `CertificateAuthenticationValidator` instance.
+
+Once this is done, the `CertificateAuthenticationFilter` can be added to the filters that will be applied to all actions:
+
+```csharp
+public void ConfigureServices(IServiceCollections services)
+{
+    services.AddScoped<ICachedSecretProvider(serviceProvider => new MyCachedSecretProvider());
```

Comment:
I think this approach deviates a bit from what we've discussed.

My understanding is that we will have a `CertificateAuthenticationConfigurationBuilder` that allows us to build the config of what to validate, for example:
```csharp
var certificateAuthenticationConfig = CertificateAuthenticationValidator
                                          .WithSubject(X509InputStore.Configuration, "my-config-key") // X509InputStore is a terrible name
                                          .WithThumbprint(X509InputStore.SecretProvider, "my-secret-name")
                                          .Build()
```

Which is then passed to the Filter to be used to run your validation against with the following configuration/criteria
```csharp
services.AddMvc(options  =>
   options.Filters.Add(new CertificateAuthenticationFilter(certificateAuthenticationConfig));
```

`CertificateAuthenticationRequirementsBuilder` is also fine for me and maybe is more explicit but the idea is the outcome of the builder is a single object, for example `CertificateAuthenticationRequirement`, which contains everything needed by the validator to run and know where to pull info from.

Did you have the same understanding @fgheysels?

Reply:
Ah, yes; now I understand it more. I didn't was quite a fan of how it is now set-up because the configuration is splintered across the application.

Reply:
But then we would have a fluent member `.With...` for each enum member. Isn't that what duplication? The alternative is passing allong, but that wouldn't probably make thinks easier.

Reply:
We would indeed have a `.With` for every enum value, but the consumers will not even know about that enumeration value given they will only use the builder or config directly.

Reply:
Yep, is now current approach.